Adding AlarmDecoder platform

hawk259 opened the pull request:
## Description:

Adding AlarmDecoder devices support.

AlarmDecoder is a USB or Raspberry PI connected device to control and monitor DSC and Ademco / Honeywell Vista alarm systems.

https://www.alarmdecoder.com

I based this off the envisalink code, big thank you to @pvizeli for that.

This has support for both SocketDevice and SerialDevice, but I don’t have access to a USB version for testing.  I left a spot for it to be added.

This adds:

alarmdevice component
alarm_control_panel shows status and allows arm/disarm
binary_sensor for all zones configured
sensor for a panel display, shows the text from the keypad display. This is off by default, enabled via the config

It requires alarmdecoder 0.12.1.0 to talk to the device.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#2373

## Example entry for `configuration.yaml` (if applicable):
```yaml
alarmdecoder:
  device:
    type: socket
    host: 10.0.0.2
    port: 10000
  panel_display: On
  zones:
    01:
      name: 'Smoke Detector'
      type: 'smoke'
    02:
      name: 'Front Door'
      type: 'opening'
    03:
      name: 'Family Room Motion'
      type: 'motion'

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

hawk259 committed:
Added AlarmDecoder platform

hawk259 committed:
remove try/catch for generic execption

gurumitts commented:
I pulled these changes into my local and the functionality works as expected.  Looks great!  I am looking forward to using this soon.  
Note:. I only test this with the socket configuration.  My alarmdecoder is on the network connected to a dedicated raspberry pi.

pvizeli commented on homeassistant/components/alarmdecoder.py:
```diff
@@ -0,0 +1,159 @@
+    vol.Optional(CONF_DEVICE_PORT, default=DEFAULT_DEVICE_PORT): cv.port,
+    vol.Optional(CONF_DEVICE_PATH, default=DEFAULT_DEVICE_PATH): cv.string,
+    vol.Optional(CONF_DEVICE_BAUD, default=DEFAULT_DEVICE_BAUD): cv.string})
+
```
Please split you validate into 3 parts (sock, serial, usb) and use vol.Any() to validate the right schema.

pvizeli reviewed (changes requested):
Look good. Can you also change the private `_` variable präfix inside normal function to normal handling as we have in all other component?

hawk259 committed:
Changes for @pvizeli, thanks for the review!

Removed _ prefix from normal function variables
Removed _hass as it will be set via .hass for us
Broke out the three config (socket, serial, usb) and use vol.Any
Added support for USB I think, don't have device, but should work
Removed components dictionary, was form old group all code that didn't work

JuliusLedoux commented:
Hello There, is this complete? When is going to be added to the official branch? Thanks!

pvizeli committed:
Fix hass string handling
